### PR TITLE
Add pipeline for Debian PR testing and Debian package release

### PR DESCRIPTION
**Pull request by ehelms**

### CODE REVIEW

**Comment by ehelms** [comment]
Contains https://github.com/theforeman/jenkins-jobs/pull/97 which should be merged first to make it easier to read this diff

File: theforeman.org/pipelines/lib/packaging.groovy
Comment (ehelms) [inline theforeman.org/pipelines/lib/packaging.groovy:153]: This depends on the outcome of https://github.com/theforeman/jenkins-jobs/pull/95/files and https://github.com/theforeman/foreman-infra/pull/1642

File: theforeman.org/pipelines/lib/packaging.groovy
Comment (evgeni) [inline theforeman.org/pipelines/lib/packaging.groovy:241]: This can be simplified if we merge https://github.com/theforeman/jenkins-jobs/pull/90 first :)

File: theforeman.org/pipelines/lib/packaging.groovy
Comment (evgeni) [inline theforeman.org/pipelines/lib/packaging.groovy:250]: same, https://github.com/theforeman/jenkins-jobs/pull/90 makes it easier

**Comment by evgeni** [comment]
#97 is merged

File: theforeman.org/pipelines/lib/packaging.groovy
Comment (ekohl) [inline theforeman.org/pipelines/lib/packaging.groovy:405]: Should this be `pdebuild` instead of `pdebuild`?

File: theforeman.org/pipelines/lib/packaging.groovy
Comment (ekohl) [inline theforeman.org/pipelines/lib/packaging.groovy:197]: I'm wondering if we could use a native method like [copyArtifacts](https://jenkinsci.github.io/job-dsl-plugin/#method/javaposse.jobdsl.dsl.helpers.step.StepContext.copyArtifacts).

File: theforeman.org/pipelines/release/foreman_packaging_deb_release.groovy
Comment (ekohl) [inline theforeman.org/pipelines/release/foreman_packaging_deb_release.groovy:73]: I don't think this should be done in parallel. We rsync them over and then rebuild the archive. That's quite an expensive operation and if you can do it just once for multiple packages that's much cheaper.

Comment (ehelms) [reply theforeman.org/pipelines/release/foreman_packaging_deb_release.groovy:73]: We already do this in parallel as I understand things. The original stage packages script is run per project-os combination. So a job like https://ci.theforeman.org/job/packaging_build_deb_dependency/559/console would run that script three times which could all happen at the same time? 

What rebuilds the archive? the rsync trigger that?

Comment (ekohl) [reply theforeman.org/pipelines/release/foreman_packaging_deb_release.groovy:73]: On the server side there's a script that runs when rsync runs:
https://github.com/theforeman/foreman-infra/blob/master/puppet/modules/freight/templates/rsync.erb

As you can see, it runs the original command and then calls
```erb
  find <%= @home %>/rsync_cache/$DISTRO/$REPO -iname '*.deb' -exec freight-add -v -c <%= @home %>/freight.conf {} apt/$DISTRO/$REPO \;
  # Publish the debs
  freight-cache -c <%= @home %>/freight.conf -v apt/$DISTRO
  # Cleanup - no need to keep the debs
  find <%= @home %>/rsync_cache/$DISTRO/$REPO -iname '*.deb' -delete || true
```

Comment (ehelms) [reply theforeman.org/pipelines/release/foreman_packaging_deb_release.groovy:73]: So what matters is `distro` and `repo`, so we could run this in parallel per OS and repository if we collected all those possible combinations up and constructed the run based on that?

Comment (ehelms) [reply theforeman.org/pipelines/release/foreman_packaging_deb_release.groovy:73]: Sorry, what I meant to ask was, is there a model where we could initiate this script directly for distro/repo pairings based on what was built given that the rsync is based upon the actual projects that are built?

Comment (ekohl) [reply theforeman.org/pipelines/release/foreman_packaging_deb_release.groovy:73]: Yes, we essentially need to gather all changed debs. Then make 2 lists:
* Those to staging (core + dependency)
* Those to release (plugin)

Then roughly `if (to_staging.any?) { rsync_to_staging(to_staging) }` and the same with `to_release`.

You're right that this is an optimization we didn't do before, but can do now.

Comment (ehelms) [reply theforeman.org/pipelines/release/foreman_packaging_deb_release.groovy:73]: Is there anything that would prevent us from moving to a model where all of or parts of https://github.com/theforeman/foreman-infra/blob/master/puppet/modules/freight/templates/rsync.erb are initiated explicitly rather than implicitly due to an rsync? I am thinking that would allow us to have cleaner optimized code if the archive building is the expensive operation.

**Comment by ehelms** [comment]
There is a bit to digest here, and where it made sense I have tried to break things out from shell into native groovy methods of code handling. The code is still a bit complex due to:

 * source generation requires different methods based on a few criteria
 * plugins are handled special in almost every case making for a lot of variation
    - their directory structure
    - that they are not staged unless its a pull request
  
I wonder if there are further optimizations we could make in the process to make it smoother to understand the branches. For example, choosing to stage plugins rather than release them directly to reduce those code branches. 

These are all things that could be done before hand or after this is merged. Given this provides various benefits to get merged (e.g. an auto release job).

File: theforeman.org/pipelines/lib/packaging.groovy
Comment (ehelms) [inline theforeman.org/pipelines/lib/packaging.groovy:368]: Attempting to make this change here first to reduce the diff https://github.com/theforeman/jenkins-jobs/pull/100

File: theforeman.org/pipelines/lib/packaging.groovy
Comment (ehelms) [inline theforeman.org/pipelines/lib/packaging.groovy:291]: I am wondering if we should pull `gem2deb` to a stand-alone repository (or foreman-infra) and deploy it to the Jenkins nodes?

File: theforeman.org/pipelines/lib/packaging.groovy
Comment (ehelms) [inline theforeman.org/pipelines/lib/packaging.groovy:341]: Anybody know why for plugins that are gems we run this download_gems script and for other gems (like dependencies) we use `gem fetch` and `gem2deb` ?

File: theforeman.org/pipelines/lib/packaging.groovy
Comment (evgeni) [reply theforeman.org/pipelines/lib/packaging.groovy:291]: Maybe. The crux behind *this* `gem2deb` is: it's a patched version :/

(not saying we can make deployment of it easier, just that we need _that_ version)

File: theforeman.org/pipelines/lib/packaging.groovy
Comment (evgeni) [inline theforeman.org/pipelines/lib/packaging.groovy:303]: we already have `archive_git_hash` in `git.groovy`, which does `git rev-parse ${ref}` (and that's equal to `rev-list and then getting the first entry)

File: theforeman.org/pipelines/lib/packaging.groovy
Comment (evgeni) [inline theforeman.org/pipelines/lib/packaging.groovy:316]: doesn't `dir(…)` do that for us already?

File: theforeman.org/pipelines/lib/packaging.groovy
Comment (evgeni) [inline theforeman.org/pipelines/lib/packaging.groovy:329]: we have similar (but not identical) versions (heh) of this in various places, and I think this is the most correct one, so how about

```groovy
def package_version(changelog_path) {
    return sh(returnStdout: true, script: "head -n1 ${changelog_path} |awk '{print \$2}'|sed 's/(//;s/)//'|cut -f1 -d-|cut -d: -f2").trim()
```

and then using it everywhere

File: theforeman.org/pipelines/lib/packaging.groovy
Comment (evgeni) [reply theforeman.org/pipelines/lib/packaging.groovy:341]: I think (and I can be totally wrong) this is due to the totally different ways of how we build/vendor stuff.
* dependencies are "real" (mostly policy compliant) Debian packages with Ruby modules built against the system Ruby, so the workflow is: use `gem fetch` to obtain the `.gem` file, call `gem2deb` on it to convert it to a Debian (source) package, build it.
* core is a big bundle (generated by `bundle package`), with some system level dependencies from above.
* plugins are *extending* this (the core) bundle, to be able to do this, we pull down the gems (of the plugin and its dependencies, using `download_gems`) and place those in the package into the same folder the core bundle is installed (and a bundler.d snippet). on installation of the package, we call `bundle update` and thus actually add those modules to the bundle.

hope this helps, and didn't confuse you further?

File: theforeman.org/pipelines/lib/packaging.groovy
Comment (evgeni) [inline theforeman.org/pipelines/lib/packaging.groovy:393]:
```suggestion
            FOREMAN_VERSION=${version} pdebuild-${os}64
```

https://github.com/theforeman/jenkins-jobs/pull/99 changed that.

File: theforeman.org/pipelines/lib/packaging.groovy
Comment (evgeni) [inline theforeman.org/pipelines/lib/packaging.groovy:396]: I *think* this is not needed anymore after #99, but I am not 100% sure.

File: theforeman.org/pipelines/lib/packaging.groovy
Comment (ekohl) [reply theforeman.org/pipelines/lib/packaging.groovy:341]: I think only Foreman plugins use bundler that way. Smart Proxy plugins are more policy compliant and AFAIK also use system Ruby (and system gems).

Comment (evgeni) [reply theforeman.org/pipelines/lib/packaging.groovy:341]: That's of course correct and makes me no more happier ;)

File: theforeman.org/pipelines/lib/packaging.groovy
Comment (ehelms) [reply theforeman.org/pipelines/lib/packaging.groovy:291]: Yea... thats why I thought about pulling it to some small dedicated repo since it won't change much or drop it in `puppet/slave/files` and deploy it to the nodes directly.

**Comment by ehelms** [comment]
Pushed another round of updates, I think the things being discussed currently here can happen after this goes in. I am trying to balance improvements with getting a release job available to reduce overhead with debian packaging.

File: theforeman.org/pipelines/lib/packaging.groovy
Comment (ehelms) [inline theforeman.org/pipelines/lib/packaging.groovy:387]: @evgeni so what should this be now?

Comment (evgeni) [reply theforeman.org/pipelines/lib/packaging.groovy:387]: sudo, sorry!

**Comment by ehelms** [comment]
There are a couple options we could choose to deploy this and get comfortable with it:

 * Deploy just the PR testing job alongside the current PR testing job and let them run together, wait on enabling the release job
 * Replace current PR testing job with this one and let it run for a it and then enable the release job
 * Turn on just the release job to test the new code and leave existing PR testing / jobs in place and then turn on PR testing later
 * Just do it: turn everything new on and remove the old and fix it as we go

File: theforeman.org/pipelines/lib/packaging.groovy
Comment (ehelms) [reply theforeman.org/pipelines/lib/packaging.groovy]: This now contains the fix from #100

File: theforeman.org/pipelines/lib/git.groovy
Comment (evgeni) [inline theforeman.org/pipelines/lib/git.groovy:20]:
```suggestion
    writeFile('commit', hash)
```

File: theforeman.org/pipelines/lib/packaging.groovy
Comment (evgeni) [inline theforeman.org/pipelines/lib/packaging.groovy:40]: what's the idea behind noting them down individually, and then returning all together? ain't it "easier" (to the eyes) to `def changed_debs = []`, then in each `for …` do `changed_debs.add(…)` and later just `return changed_debs`?

File: theforeman.org/pipelines/lib/packaging.groovy
Comment (evgeni) [inline theforeman.org/pipelines/lib/packaging.groovy:133]: this is never used again?

File: theforeman.org/pipelines/lib/packaging.groovy
Comment (evgeni) [inline theforeman.org/pipelines/lib/packaging.groovy:150]:
```suggestion
                release_type = 'release'
                echo "plugin build without PR: uploading directly to deb/${suite}/${component}"
```

File: theforeman.org/pipelines/lib/packaging.groovy
Comment (evgeni) [inline theforeman.org/pipelines/lib/packaging.groovy:166]:
```suggestion
            repos["${suite}-${component}"].deb_paths.add("${deb_path}/*deb")
```

File: theforeman.org/pipelines/lib/packaging.groovy
Comment (evgeni) [inline theforeman.org/pipelines/lib/packaging.groovy:225]: why `=~` and not `=`? also this should be `nightly`, not `develop`, for the same reasons as in https://github.com/theforeman/jenkins-jobs/pull/96/files#r700835168

File: theforeman.org/pipelines/release/foreman_packaging_deb_release.groovy
Comment (evgeni) [inline theforeman.org/pipelines/release/foreman_packaging_deb_release.groovy:47]:
```suggestion
                        build_steps = []
```

Comment (evgeni) [reply theforeman.org/pipelines/release/foreman_packaging_deb_release.groovy:47]: or whatever is the correct syntax, to make the other steps be noop. probably `build_steps = ['setup_sources': [], 'executes': [], 'rsync_packages': []]` or so?

File: theforeman.org/pipelines/test/deb_packaging.groovy
Comment (evgeni) [inline theforeman.org/pipelines/test/deb_packaging.groovy:31]: is there a case where the *test* pipeline is triggered without a PR? you do rely on `ghprbPullAuthorLogin`, so IMHO we can just assume that `pull_request=true`.

File: theforeman.org/scripts/debian/build_project.sh
Comment (evgeni) [inline theforeman.org/scripts/debian/build_project.sh:1]: what is this file for?

File: theforeman.org/pipelines/lib/packaging.groovy
Comment (evgeni) [inline theforeman.org/pipelines/lib/packaging.groovy:250]:
```suggestion
            tar xf ${project}_${package_version}.orig.tar.bz2
```

* `j` is not needed as `tar` can figure out the compression on its own
* `v` makes the build log so noisy, that I'd prefer if we would drop it

File: theforeman.org/pipelines/lib/git.groovy
Comment (ehelms) [reply theforeman.org/pipelines/lib/git.groovy:20]: huh, wth was I thinking

File: theforeman.org/pipelines/lib/packaging.groovy
Comment (ehelms) [reply theforeman.org/pipelines/lib/packaging.groovy:225]: Holdovers from trying to translate shell scripts to groovy. Thanks for the catch.

File: theforeman.org/pipelines/test/deb_packaging.groovy
Comment (ehelms) [reply theforeman.org/pipelines/test/deb_packaging.groovy:31]: Yes, that is true now. When I wrote this I was still potentially re-using this for the release pipeline.

File: theforeman.org/scripts/debian/build_project.sh
Comment (ehelms) [reply theforeman.org/scripts/debian/build_project.sh:1]: Absolutely nothing. Was accidental add from when I thought about trying to create a consolidated build script that combined all the ways to build source and then realized how crazy it is.

File: theforeman.org/pipelines/lib/git.groovy
Comment (evgeni) [reply theforeman.org/pipelines/lib/git.groovy]: dash is the default `/bin/sh` on Debian, so idk?

File: theforeman.org/pipelines/lib/packaging.groovy
Comment (evgeni) [inline theforeman.org/pipelines/lib/packaging.groovy:45]: I think you can use something link this:

```suggestion
        (folder, os, project) = core.split('/')
```

and then use the better names below.

```console
groovy:000> s = "debian/focal/foreman/changelog"
===> debian/focal/foreman/changelog
groovy:000> (type, os, project) = s.split('/')
===> [Ljava.lang.String;@dc7df28
groovy:000> type
===> debian
groovy:000> project
===> foreman
groovy:000> os
===> focal
```

File: theforeman.org/pipelines/lib/packaging.groovy
Comment (evgeni) [inline theforeman.org/pipelines/lib/packaging.groovy:129]: looking at `find_changed_debs`, I don't see how `operating_systems` can be longer than 1, as you add a new entry for every changed changelog, instead of updating a possibly existing one.

File: theforeman.org/pipelines/lib/packaging.groovy
Comment (evgeni) [inline theforeman.org/pipelines/lib/packaging.groovy:62]:
```suggestion
            operating_systems: [dependency_split[1]]
```

otherwise this will be a string, not an array?

File: theforeman.org/pipelines/lib/packaging.groovy
Comment (ehelms) [reply theforeman.org/pipelines/lib/packaging.groovy:129]: Good catch, that logic changed so many times I missed this simplification

**Comment by evgeni** [comment]
There is a syntax error somewhere, as all release related pipelines produce the following error:

```
{'error': "expecting ')', found '=' @ line 446, column 21."}
output/foreman-2.3-release-pipeline/config.xml: NOT OK
```

File: theforeman.org/pipelines/lib/packaging.groovy
Comment (evgeni) [inline theforeman.org/pipelines/lib/packaging.groovy:219]:
```suggestion
        if (version == 'nightly') {
```

makes the linter so much happier ;)

File: theforeman.org/pipelines/lib/packaging.groovy
Comment (evgeni) [inline theforeman.org/pipelines/lib/packaging.groovy:45]: in https://github.com/theforeman/jenkins-jobs/pull/96#discussion_r701655615, I used `(folder, os, project)`, as it felt like using `type` here will welcome someone to refactor the line 3 down (`type: 'core'`) to read `type: type` and that'd break things.

not blocking on this, but I think not calling this thing "type" is cleaner.

Comment (ehelms) [reply theforeman.org/pipelines/lib/packaging.groovy]: type type type

**Review by evgeni — APPROVED** [review]
there are still too many shell blocks, but I think we can tackle those one by one.

**Comment by ehelms** [comment]
This now disables the current PR test and adds but keeps disabled the new release job. I am aiming to merge this on 9/7 to actively test Debian PR. Hopefully there will be a few open to try this out on.

**Pull request merged**
